packages/@uppy/status-bar/src/calculateProcessingProgress.ts

```diff
@@ -22,5 +27,5 @@ export default function calculateProcessingProgress (files) {
     mode,
     message,
     value,
-  }
+  } as FileProcessingInfo
```

Comment:
I think it's better to avoid `as` when possible, and instead do:

```
const fpi: FileProcessingInfo = {
    mode,
    message,
    value,
  }

return fpi
```

Reply:
I think having an intermediary variable is wasteful, I'd rather use `as`

Reply:
I think `as` operator should be avoided whenever possible, because it disables type checking and blindly casts the object, making it easy to make mistakes when changing the literal or its types in the future, see https://www.reddit.com/r/typescript/comments/z8f7mf/are_there_common_bad_usages_of_as/iybe027/

Reply:
```
packages/@uppy/status-bar/src/calculateProcessingProgress.ts:26:9 - error TS2322: Type '{ mode: "determinate" | "indeterminate"; message: string | undefined; value: number; }' is not assignable to type 'FileProcessingInfo'.
  Types of property 'mode' are incompatible.
    Type '"determinate" | "indeterminate"' is not assignable to type '"determinate"'.
      Type '"indeterminate"' is not assignable to type '"determinate"'.

26   const fpi: FileProcessingInfo = {
```

Reply:
that means there's probably something wrong with our types and it should be fixed either by fixing the type or adding a runtime check

Reply:
My assessment is that there's something wrong with TS 😅 I think TS is a tool, supposed to help us. If it's unable to, we should not be afraid to teach it that we are right, and it is wrong, like here. Otherwise this transition will never be completed.

Reply:
> it disables type checking and blindly casts the object

I don't think that's true, TS would complain if you try to cast to an incompatible type: https://www.typescriptlang.org/play?#code/NoRgugBAhgzhBGB7RAbAplAdsMAoIA
Using `as` can be a way to make your code stricter sometime (e.g. `as const`)

Reply:
I think it's better to not lie to typescript and instead do something like this:

```js
  if (mode === 'determinate' && message != null) {
    return {
      mode,
      message,
      value,
    }
  }

  return {
    mode: 'indeterminate',
    message: undefined,
    value: 0,
  }
```

Reply:
We're not lying though, when `mode` is `'indeterminate'`, `message` value _is_ `undefined` and `value` value _is_ `0`. TS is not able to assess that because it's not smart enough.

Reply:
> `message` value _is_ `undefined` and `value` value _is_ `0`

right now it is, yes. but if we refactor this in the future, that might no longer be the case, but typescript won't know it because we use the `as` keyword to lie to typescript